File: 1081-smallest-subsequence-of-distinct-characters/solution.py
class Solution:
    def smallestSubsequence(self, s: str) -> str:
        # Map each character to its last occurring index
        last_occurrence = {c: i for i, c in enumerate(s)}
        
        stack = []
        visited = set()
        
        for i, char in enumerate(s):
            # If the character is already in the stack, skip it
            if char in visited:
                continue
            
            # While stack is not empty, current char is smaller than top,
            # and the top character appears later in the string
            while stack and char < stack[-1] and last_occurrence[stack[-1]] > i:
                top = stack.pop()
                visited.remove(top)
            
            stack.append(char)
            visited.add(char)
            
        return "".join(stack)

File: 1081-smallest-subsequence-of-distinct-characters/solution.cpp
#include <string>
#include <vector>
#include <unordered_map>

class Solution {
public:
    std::string smallestSubsequence(std::string s) {
        std::vector<int> last_occurrence(26, -1);
        for (int i = 0; i < s.length(); ++i) {
            last_occurrence[s[i] - 'a'] = i;
        }

        std::vector<bool> visited(26, false);
        std::string stack = "";

        for (int i = 0; i < s.length(); ++i) {
            char c = s[i];
            if (visited[c - 'a']) continue;

            // While stack is not empty, current char is smaller than top,
            // and the top character appears later in the string
            while (!stack.empty() && c < stack.back() && last_occurrence[stack.back() - 'a'] > i) {
                visited[stack.back() - 'a'] = false;
                stack.pop_back();
            }

            stack.push_back(c);
            visited[c - 'a'] = true;
        }

        return stack;
    }
};